[ARMv7] Fix OMG Int64 live value OSR restore

justinmichaud opened the pull request:
#### d2b0596d08f44b20f0a488e7b6ccb4c03af36ee3
<pre>
[ARMv7] Fix OMG Int64 live value OSR restore
<a href="https://bugs.webkit.org/show_bug.cgi?id=282309">https://bugs.webkit.org/show_bug.cgi?id=282309</a>

Reviewed by David Degazio.

Int64 patchpoints get lowered in a way that makes the indices too hard
to compute when filling live values from patchpoints during exception unwinding.

1) Handle patchpoint live-value saving directly in the generator by splitting Int64s.
2) Fix BBQ OSR, which needs a value rep for register pairs. That is, LLInt fills
Int64 values directly from the stack. BBQ may map them to/from a register pair,
but still saves them to the OMG OSR buffer as a 64-bit value. OMG does an Int64
load, which is agressively lowered to remove Int64 stitches / extracts.

We also add some tests to make debugging OSR very simple.

* JSTests/wasm/stress/armv7-simple-throw-restore-live.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.tag.e.func.test.export.string_appeared_here.param.a.i32.result.i64.local.x.i64.local.x.i64.const.0x0EEFBAADDEADFEED.try.local.x.i64.add.i64.const.1.local.x.call.fun.local.a.local.x.i64.add.i64.const.1.local.x.catch.e.local.x.i64.add.i64.const.5.local.x.return.local.x.return.local.x.func.fun.param.x.i32.i32.eq.local.x.i32.const.42.then.throw.e.async test):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::traceCF):
(JSC::Wasm::OMGIRGenerator::loadFromScratchBuffer):
(JSC::Wasm::OMGIRGenerator::preparePatchpointForExceptions):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::loadValuesIntoBuffer):

Canonical link: <a href="https://commits.webkit.org/286405@main">https://commits.webkit.org/286405@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c605dd3580f298c504cbc2fed5f7f5c40010bc3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75859 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54888 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28756 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80356 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27125 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77975 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64030 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3182 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59481 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17642 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78926 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49357 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65151 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39841 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46757 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22634 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25452 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69036 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67872 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22972 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81820 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75135 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3228 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2028 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67710 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3382 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65119 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67016 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16709 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10960 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9086 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97389 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3178 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21293 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3199 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4137 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3206 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aoikonomopoulos commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -4349,8 +4349,15 @@ auto OMGIRGenerator::addSIMDLoadPad(SIMDLaneOperation op, ExpressionType pointer
+        auto* lo = append<MemoryValue>(m_proc, Load, Int32, origin(), pointer, offset);
+        auto* hi = append<MemoryValue>(m_proc, Load, Int32, origin(), pointer, offsetHi);
+        return append<Value>(m_proc, Stitch, type, origin(), lo, hi);
+    }
```
Hmm, I don't immediately see how this can work when `loadFromScratchBuffer` is used in the OSR path? Doesn't this introduce a mismatch between the values that the lower tier saves and the values that we restore (i.e. we expect Int64s broken up in two parts and stored in two 64-bit slots) but the lower tier will just store an Int64 in a single 64-bit slot?

justinmichaud replied:
Yeah, I missed this. I added a test

ddegazio commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -2354,6 +2356,83 @@ JSC_DEFINE_HOST_FUNCTION(functionFTLTrue, (JSGlobalObject*, CallFrame*))
+        if (visitor->codeType() != StackVisitor::Frame::Wasm
+            || !visitor->callee().isNativeCallee()) {
+            allFramesAreValid = false;
+            dataLogLn("C");
```
I'm guessing this datalog isn't meant to be committed?

justinmichaud replied:
I think it is helpful if the test is failing to quickly be able to tell why, but it isn't strictly needed.

ddegazio commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -2354,6 +2356,83 @@ JSC_DEFINE_HOST_FUNCTION(functionFTLTrue, (JSGlobalObject*, CallFrame*))
+        if (visitor->codeType() != StackVisitor::Frame::Wasm
+            || !visitor->callee().isNativeCallee()
+            || !isAnyWasmToJS(static_cast<Wasm::Callee*>(visitor->callee().asNativeCallee())->compilationMode())) {
+            dataLogLn("B");
```
Ditto.

ddegazio commented on Source/JavaScriptCore/tools/JSDollarVM.cpp:
```diff
@@ -2354,6 +2356,83 @@ JSC_DEFINE_HOST_FUNCTION(functionFTLTrue, (JSGlobalObject*, CallFrame*))
+    auto expectNative = [&](StackVisitor& visitor) {
+        if (visitor->codeType() != StackVisitor::Frame::Native
+            || !visitor->callee().isCell()) {
+            dataLogLn("A");
```
Ditto.

ddegazio commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
```diff
@@ -1013,8 +1013,13 @@ class OMGIRGenerator {
-    Value* loadFromScratchBuffer(unsigned& indexInBuffer, Value* pointer, B3::Type);
-    void connectControlAtEntrypoint(unsigned& indexInBuffer, Value* pointer, ControlData&, Stack& expressionStack, ControlData& currentData, bool fillLoopPhis = false);
+    enum OSRBufferMode {
+        LoadI64, // Load I64 values direclty, and let later passes lower them; Needed for interacting with lower tiers.
```
Typo, `direclty` -> `directly`.

ddegazio reviewed (approved):
r=me with comments.

justinmichaud committed:
[ARMv7] Fix OMG Int64 live value OSR restore
https://bugs.webkit.org/show_bug.cgi?id=282309

Reviewed by David Degazio.

Int64 patchpoints get lowered in a way that makes the indices too hard
to compute when filling live values from patchpoints during exception unwinding.

1) Handle patchpoint live-value saving directly in the generator by splitting Int64s.
2) Fix BBQ OSR, which needs a value rep for register pairs. That is, LLInt fills
Int64 values directly from the stack. BBQ may map them to/from a register pair,
but still saves them to the OMG OSR buffer as a 64-bit value. OMG does an Int64
load, which is agressively lowered to remove Int64 stitches / extracts.

We also add some tests to make debugging OSR very simple.

* JSTests/wasm/stress/armv7-simple-throw-restore-live.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.tag.e.func.test.export.string_appeared_here.param.a.i32.result.i64.local.x.i64.local.x.i64.const.0x0EEFBAADDEADFEED.try.local.x.i64.add.i64.const.1.local.x.call.fun.local.a.local.x.i64.add.i64.const.1.local.x.catch.e.local.x.i64.add.i64.const.5.local.x.return.local.x.return.local.x.func.fun.param.x.i32.i32.eq.local.x.i32.const.42.then.throw.e.async test):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator32_64.cpp:
(JSC::Wasm::OMGIRGenerator::traceCF):
(JSC::Wasm::OMGIRGenerator::loadFromScratchBuffer):
(JSC::Wasm::OMGIRGenerator::preparePatchpointForExceptions):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::loadValuesIntoBuffer):

Canonical link: https://commits.webkit.org/286405@main
